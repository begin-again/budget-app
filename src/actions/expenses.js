import database from '../firebase/firebase';

export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({ id: ref.key, ...expense }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

/**
 *
 * @param {String} id
 * @param {Object} updates
 */
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// export const startSetExpenses;
// 1. fetch all expense data
// 2. Parse into and array
// 3. dispatch SET_EXPENSES
export const startSetExpenses = () => {
  return (dispatch) => {
    return database
      .ref('expenses')
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(childSnap => {
          expenses.push({ id: childSnap.key, ...childSnap.val() });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
