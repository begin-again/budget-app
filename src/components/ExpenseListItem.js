import React from 'react';
import { Link } from 'react-router-dom';

export default ({ description, amount, createdAt, id }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>
        {description}
      </h3>
    </Link>
    <p>{amount} &raquo; {createdAt}</p>
  </div>
);
