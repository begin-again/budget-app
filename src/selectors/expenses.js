
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense, i) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      let textMatch = true;
      if (text) {
        const rx = new RegExp(text, 'i');
        textMatch = rx.test(expense.description);
      }

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') return b.createdAt - a.createdAt;
      return b.amount - a.amount;
    });
};
