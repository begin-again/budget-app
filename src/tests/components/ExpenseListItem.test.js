/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

describe('Expense List Item Component', () => {
  it('should render', () => {
    const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);

    expect(wrapper).toMatchSnapshot();
  });
});
