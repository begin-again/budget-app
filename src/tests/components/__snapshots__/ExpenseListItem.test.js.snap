// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense List Item Component should render 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      sweet gum
    </h3>
  </Link>
  <p>
    $1.95
    <span
      className="seperator"
    >
      Â»
    </span>
    Dec 31st 1969
  </p>
</div>
`;
